### PR TITLE
Bring vip-caching.php up to date with WordPress.com's wpcom-caching.php

### PR DESCRIPTION
**Pull request by emrikol**
I've done this in several commits to try and keep parity with the changes made on WordPress.com

### CODE REVIEW

**Comment by joshbetz** [comment]
Supersedes #203

File: vip-helpers/vip-caching.php [hunk 1]
```diff
@@ -188,7 +188,11 @@ function wpcom_vip_get_page_by_path( $page_path, $output = OBJECT, $post_type =
 	if ( $page_id === false ) {
 		$page = get_page_by_path( $page_path, $output, $post_type );
 		$page_id = $page ? $page->ID : 0;
-		wp_cache_set( $cache_key, $page_id, 'get_page_by_path' ); // We only store the ID to keep our footprint small
+		if ( $page_id ===0 ){
```
Comment (joshbetz) [inline vip-helpers/vip-caching.php]: Minor nit: spacing issue here between `===0`

Comment (emrikol) [inline vip-helpers/vip-caching.php]: This is a direct copy from WordPress.com SVN.  I'd be happy to send it through PHPCS and have it auto-fix whitespace issues.

File: vip-helpers/vip-caching.php [hunk 2]
```diff
@@ -391,6 +391,78 @@ function wpcom_vip_get_nav_menu_object( $menu ) {
...
 
 /**
+ * Require the Stampedeless_Cache class for use in our helper functions below.
+ *
+ * The Stampedeless_Cache helps prevent cache stampedes by internally varying the cache
+ * expiration slightly when creating a cache entry in an effort to avoid multiple keys
+ * expiring simultaneously and allowing a single request to regenerate the cache shortly
+ * before it's expiration.
+ */
+if( function_exists( 'require_lib' ) && defined( 'WPCOM_IS_VIP_ENV' ) && WPCOM_IS_VIP_ENV )
```
Comment (joshbetz) [inline vip-helpers/vip-caching.php]: We should add curly braces for this conditional

**Comment by joshbetz** [comment]
@emrikol Only a couple minor issues. I'm going to merge, but it would be cool to fix them anyway if you have time.

**Comment by mjangda** [comment]
r85563-deploy

**Pull request merged**